src/segment_reshape/topology/find_related.py

```diff
@@ -328,4 +315,78 @@ def get_common_geometries(
                 )
             )
 
-    return segment, common_parts, edges
+    return CommonGeometriesResult(segment, common_parts, edges)
+
+
+def _as_line_segments(geometry: QgsGeometry) -> Set[Segment]:
+    vertices = [
+        (point.x(), point.y()) for point in cast(QgsLineString, geometry.asPolyline())
+    ]
+    return {frozenset(line) for line in zip(vertices, vertices[1:])}
+
+
+def _build_line_from_line_segment_set(
+    trigger_part: QgsGeometry,
+    line_segments_to_keep: Set[Segment],
+    trigger_segment: Segment,
+    edge_candidate_geometries: List[QgsGeometry],
+) -> QgsLineString:
+    """Build a subline from trigger_part with constraints
+
+    Builds a line which contains the trigger_segment and has only segments from
+    line_segments_to_keep set and which is split at points where any geometry
+    from edge_candidate_geometries touches the line.
+
+    Args:
+        trigger_part (QgsGeometry): Input geometry from where the subline is extracted
+        line_segments_to_keep (Set[Segment]): Set of segments that the subline is build
+            from
+        trigger_segment (Segment): Edge which was clicked
+        edge_candidate_geometries (List[QgsGeometry]): List of geometries which might
+            act as break points
+
+    Returns:
+        QgsLineString: A sublinestring from the input geometry that follows the
+            constraints
+    """
+
+    possible_split_points = {
+        (vertex.x(), vertex.y())
+        for geom in edge_candidate_geometries
+        for vertex in geom.vertices()
+    } & {(vertex.x(), vertex.y()) for vertex in trigger_part.vertices()}
+
+    vertex_iterator = trigger_part.vertices()
+    next_vertex_iterator = trigger_part.vertices()
+    _ = next(next_vertex_iterator)  # advance the next_vertex iterator by one
+
+    parts: List[QgsPoint] = []
+    current: List[QgsPoint] = []
```

Comment:
Could this be renamed to something more descriptive (current_points or what?)? It is quite difficult to understand its purpose especially in conditionals (`if not current:` or `if next_vertex_tuple in possible_split_points and current` -> not very readable).

Reply:
fixed